src/arithmetic/agg_funcs.c

```diff
@@ -221,6 +224,57 @@ AggCtx *Agg_CountFunc() {
 
 //------------------------------------------------------------------------
 
+typedef struct {
+	SIValue *values;
+} __agg_countDistinctCtx;
+
+int __agg_countDistinctStep(AggCtx *ctx, SIValue *argv, int argc) {
+	assert(argc == 1);
+	__agg_countDistinctCtx *ac = Agg_FuncCtx(ctx);
+	if(SIValue_IsNull(argv[0])) return AGG_OK;
+
+	ac->values = array_append(ac->values, SI_CloneValue(argv[0]));
+	return AGG_OK;
+}
+
+int __agg_countDistinctReduceNext(AggCtx *ctx) {
+	__agg_countDistinctCtx *ac = Agg_FuncCtx(ctx);
+
+	uint32_t item_count = array_len(ac->values);
+	if(item_count <= 1) {
+		Agg_SetResult(ctx, SI_LongVal(item_count));
+		return AGG_OK;
+	}
+
+	/* Sort array, don't count duplicates */
+	QSORT(SIValue, ac->values, item_count, SI_LT);
```

Comment:
Wouldn't it be a lot more efficient to update a hash/rax with incoming values rather than this approach?

Reply:
What would you be using for keys?

Reply:
I think an `SIValue_Hash64` function using logic like that of `Record_Hash64` would be a solid solution! (Incidentally, we need to write array-hashing logic to update that function.)

Reply:
For hashing an array we can use something like Java's ArrayList `hashCode` function, which hashes Java's array list (vector) data type
```
public int hashCode() {
    int hashCode = 1;
    for (E e : this)
        hashCode = 31*hashCode + (e==null ? 0 : e.hashCode());
    return hashCode;
}
```
I will gladly help with implementing hashing if needed